### PR TITLE
Implement `Governance` module

### PR DESCRIPTION
**Pull request by TomTaehoonKim**

### CODE REVIEW

File: governance/src/lib.rs [hunk 1]
```diff
@@ -29,15 +29,33 @@ pub struct Governance<N: GossipNetwork, S: Storage> {
...
+                    .or_insert_with(HashSet::new)
+                    .insert(voter);
+                votes
+            });
+        let height = self.dms.read_height().await?;
+        let status = GovernanceStatus { votes, height };
+        Ok(status)
     }
 
     pub async fn vote(
```
Comment (YoonhoKim0527) [inline governance/src/lib.rs:61]: I can't accept the function name `vote` here . 
This function makes a data, makes a message from data and adds to storage and broadcast. 
Can we think about a better function name? (i.e. add_message_and_broadcast or something mean like that)

Comment (YoonhoKim0527) [inline governance/src/lib.rs:61]: I can't accept the function name `vote` here . 
This function makes a data, makes a message from data and adds to storage and broadcast. 
Can we think about a better function name? (i.e. add_message_and_broadcast or something mean like that)

Comment (TomTaehoonKim) [reply governance/src/lib.rs:61]: > This function makes a data, makes a message from data and adds to storage and broadcast.

This is exactly what `voting` is in the Governance model.
It's a higher level abstraction with regards to the DMS's abstraction. Your suggestion is more appropriate for the lower level abstraction i.e., DMS.

Comment (YoonhoKim0527) [reply governance/src/lib.rs:61]: Agree!

File: governance/src/lib.rs [hunk 2]
```diff
@@ -29,15 +29,33 @@ pub struct Governance<N: GossipNetwork, S: Storage> {
...
-    pub async fn read(&self) -> Result<GovernanceState, Error> {
-        unimplemented!()
+    pub async fn read(&self) -> Result<GovernanceStatus, Error> {
+        let messages = self.dms.read_messages().await?;
+        let votes = messages
+            .iter()
+            .map(|message| {
+                (
+                    message.data().to_string().to_hash256(),
+                    message.signature().signer().clone(),
```
Comment (junha1) [inline governance/src/lib.rs:46]: This is not the 'voter'. It's just network packet signer. See line 71 and  76

File: governance/src/lib.rs [hunk 3]
```diff
@@ -29,15 +29,33 @@ pub struct Governance<N: GossipNetwork, S: Storage> {
...
 
-    pub async fn read(&self) -> Result<GovernanceState, Error> {
-        unimplemented!()
+    pub async fn read(&self) -> Result<GovernanceStatus, Error> {
+        let messages = self.dms.read_messages().await?;
+        let votes = messages
+            .iter()
+            .map(|message| {
+                (
+                    message.data().to_string().to_hash256(),
```
Comment (junha1) [inline governance/src/lib.rs:45]: We need `agenda_hash`, not the hash of the data.

File: governance/src/lib.rs (hunk 2 above)
Comment (TomTaehoonKim) [reply governance/src/lib.rs]: fixed

File: governance/src/lib.rs (hunk 3 above)
Comment (TomTaehoonKim) [reply governance/src/lib.rs]: fixed

**Review by junha1 — APPROVED** [review]

**Pull request merged**